Allow WebKit to run main-thread frames in parallel with CoreGraphics.

mattwoodrow opened the pull request:
#### 0e347d3bd4afbe1b11ea2e18d3f23b1291bc968b
<pre>
Allow WebKit to run main-thread frames in parallel with CoreGraphics.
<a href="https://bugs.webkit.org/show_bug.cgi?id=302187">https://bugs.webkit.org/show_bug.cgi?id=302187</a>
&lt;<a href="https://rdar.apple.com/163067859">rdar://163067859</a>&gt;

Reviewed by Simon Fraser.

Rather than waiting on the GPU process to finish CPU-side flushing of
CoreGraphics before allowing the next rAF to start, allow one previous frame to
still be in progress.

Use RemoteLayerTreeDrawingAreaProxy::notifyPendingCommitLayerTree to determine
when a frame has finished the WebKit main thread work, and allow sending a
second displayDidRefresh message at this time, before the previous
commitLayerTree message has been received.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultAllowMultipleCommitLayerTreePending):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::notifyPendingCommitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::allowMultipleCommitLayerTreePending):
(WebKit::RemoteLayerTreeDrawingAreaProxy::ProcessState::canSendDisplayDidRefresh):
(WebKit::RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay):

Canonical link: <a href="https://commits.webkit.org/303024@main">https://commits.webkit.org/303024@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7af7403a31b4f2a7cf26ec58ae29193fa0d77e49

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130880 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3203 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41839 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138308 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/3241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3046 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99731 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a1e47a06-a3e0-4835-a034-af9f0a5b2c02) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133826 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2303 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117198 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80418 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/01e28a50-fabd-41d3-8e06-996f698ab734) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2221 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35327 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81565 "Built successfully") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122897 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled JSC (warnings); Running jscore-test") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110810 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-api-tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35835 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140788 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129333 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2947 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2664 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108245 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2993 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113529 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108159 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27538 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2259 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31953 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55984 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3015 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66407 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162350 "Built successfully") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2836 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40497 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3036 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2944 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/html/canvas/CanvasRenderingContext.h:
```diff
@@ -107,7 +107,7 @@ class CanvasRenderingContext : public ScriptWrappable, public CanMakeWeakPtr<Can
 
     // Called periodically if needsFlush() was true when canvas change happened.
-    virtual void flushDeferredOperations() { }
+    virtual uint64_t flushDeferredOperations() { return 0; }
```
What does the return value represent? Can it get a strong type?

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2444,9 +2444,12 @@ void Page::doAfterUpdateRendering()
+    forEachRenderableDocument([&] (Document& document) {
+        deferredCanvasOperationsFlushed += document.prepareCanvasesForDisplayOrFlushIfNeeded();
     });
+    static const uint64_t kLotsOfCanvasDrawing = 10000;
```
lolz

smfr commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2346,7 +2346,7 @@ void CanvasRenderingContext2DBase::didDraw(std::optional<FloatRect> rect, Option
         return;
 
-    m_hasDeferredOperations = true;
+    m_deferredOperations++;
```
`++ m_deferredOperations`

smfr commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
```diff
@@ -504,7 +504,7 @@ class CanvasRenderingContext2DBase : public CanvasRenderingContext, public Canva
     mutable Variant<CachedContentsTransparent, CachedContentsUnknown, CachedContentsImageData> m_cachedContents;
     CanvasRenderingContext2DSettings m_settings;
-    bool m_hasDeferredOperations { false };
+    uint64_t m_deferredOperations { 0 };
```
`m_deferredOperationCount`

smfr commented on Source/WebCore/page/Page.h:
```diff
@@ -1219,6 +1219,8 @@ class Page : public RefCountedAndCanMakeWeakPtr<Page>, public Supplementable<Pag
 
     void updateScreenSupportedContentsFormats();
 
+    bool drewLotsOfCanvasInLastUpdate() const { return m_lotsOfCanvasDrawingInLastUpdate; }
```
Could we formalize this a little bit?

mattwoodrow replied:
I've removed all of the time for the time being, will see how much it matters in practice.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2444,9 +2444,12 @@ void Page::doAfterUpdateRendering()
-        document.prepareCanvasesForDisplayOrFlushIfNeeded();
+    uint64_t deferredCanvasOperationsFlushed = 0;
+    forEachRenderableDocument([&] (Document& document) {
+        deferredCanvasOperationsFlushed += document.prepareCanvasesForDisplayOrFlushIfNeeded();
```
Summing canvas operations per document isn't going to work with site isolation.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -417,14 +424,13 @@ - (void)drawInContext:(CGContextRef)cgContext
         ProcessState& state = processStateForConnection(connection);
         auto& commitLayerTreePending = std::get<CommitLayerTreePending>(state.commitLayerTreeMessageState);
-        if (!--commitLayerTreePending.requestedCommitLayerTree) {
+        commitLayerTreePending.requestedCommitLayerTree--;
```
This needs an assertion that checks for underflow.

mattwoodrow replied:
There's a MESSAGE_CHECK at the start of this function that checks it. I guess I can assert again, to make sure nobody changed it since.

mattwoodrow committed:
Allow WebKit to run main-thread frames in parallel with CoreGraphics.
https://bugs.webkit.org/show_bug.cgi?id=302187
<rdar://163067859>

Reviewed by Simon Fraser.

Rather than waiting on the GPU process to finish CPU-side flushing of
CoreGraphics before allowing the next rAF to start, allow one previous frame to
still be in progress.

Use RemoteLayerTreeDrawingAreaProxy::notifyPendingCommitLayerTree to determine
when a frame has finished the WebKit main thread work, and allow sending a
second displayDidRefresh message at this time, before the previous
commitLayerTree message has been received.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnable.h:
* Source/WebKit/Shared/WebPreferencesDefaultValues.cpp:
(WebKit::defaultAllowMultipleCommitLayerTreePending):
* Source/WebKit/Shared/WebPreferencesDefaultValues.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::notifyPendingCommitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::allowMultipleCommitLayerTreePending):
(WebKit::RemoteLayerTreeDrawingAreaProxy::ProcessState::canSendDisplayDidRefresh):
(WebKit::RemoteLayerTreeDrawingAreaProxy::didRefreshDisplay):

Canonical link: https://commits.webkit.org/303024@main
